File: backend/app/schemas/explanation.py
"""
Pydantic schemas for Explanation API
Request and response models with validation
"""

from pydantic import BaseModel, Field, field_validator
from datetime import datetime
from typing import Optional, Literal, Dict, Any
from enum import Enum


class Language(str, Enum):
    """Supported languages for explanations"""

    EN = "en"
    FR = "fr"
    DE = "de"
    ES = "es"
    IT = "it"
    ZH = "zh"


class DetailLevel(str, Enum):
    """Content detail level"""

    BRIEF = "brief"  # 2-3 sentences
    STANDARD = "standard"  # 5-8 sentences
    DETAILED = "detailed"  # 10+ sentences


class ExplanationRequest(BaseModel):
    """Request schema for generating artwork explanation"""

    artwork_name: str = Field(..., min_length=1, max_length=255, description="Artwork name")
    language: Language = Field(default=Language.EN, description="Target language")
    recognition_id: Optional[str] = Field(None, description="Associated recognition result ID")

    # Optional parameters
    detail_level: DetailLevel = Field(
        default=DetailLevel.STANDARD, description="Content detail level"
    )
    include_audio: bool = Field(default=False, description="Generate TTS audio")

    @field_validator("artwork_name")
    @classmethod
    def validate_artwork_name(cls, v: str) -> str:
        """Validate and normalize artwork name"""
        v = v.strip()
        if not v:
            raise ValueError("Artwork name cannot be empty")
        return v

    class Config:
        json_schema_extra = {
            "example": {
                "artwork_name": "Mona Lisa",
                "language": "en",
                "detail_level": "standard",
                "include_audio": False,
            }
        }


class ExplanationResponse(BaseModel):
    """Response schema for explanation content"""

    id: str = Field(..., description="Unique explanation ID")
    artwork_name: str = Field(..., description="Artwork name")
    language: str = Field(..., description="Language code")
    content: str = Field(..., description="Explanation content")
    audio_url: Optional[str] = Field(None, description="TTS audio URL")
    audio_duration_seconds: Optional[int] = Field(None, description="Audio duration")

    # Metadata
    timestamp: datetime = Field(..., description="Creation timestamp")
    cached: bool = Field(default=False, description="Whether served from cache")
    processing_time_ms: int = Field(default=0, description="Processing time in milliseconds")
    metadata: Dict[str, Any] = Field(default_factory=dict, description="Additional metadata")

    class Config:
        from_attributes = True
        json_schema_extra = {
            "example": {
                "id": "550e8400-e29b-41d4-a716-446655440000",
                "artwork_name": "Mona Lisa",
                "language": "en",
                "content": "The Mona Lisa is a portrait painting by Leonardo da Vinci...",
                "audio_url": "https://cdn.gomuseum.com/audio/550e8400.mp3",
                "audio_duration_seconds": 45,
                "timestamp": "2024-10-03T12:00:00",
                "cached": False,
                "processing_time_ms": 1200,
            }
        }


class TTSRequest(BaseModel):
    """Request schema for generating TTS audio"""

    explanation_id: str = Field(..., description="Explanation content ID")
    voice: Literal["alloy", "echo", "fable", "onyx", "nova", "shimmer"] = Field(
        default="alloy", description="Voice selection"
    )
    speed: float = Field(default=1.0, ge=0.25, le=4.0, description="Speech speed multiplier")

    class Config:
        json_schema_extra = {
            "example": {"explanation_id": "550e8400-e29b-41d4-a716-446655440000", "voice": "alloy", "speed": 1.0}
        }


class TTSResponse(BaseModel):
    """Response schema for TTS audio"""

    audio_url: str = Field(..., description="Audio file URL")
    duration_seconds: int = Field(..., description="Audio duration")
    file_size_bytes: int = Field(..., description="File size")
    cached: bool = Field(default=False, description="Whether served from cache")

    class Config:
        json_schema_extra = {
            "example": {
                "audio_url": "https://cdn.gomuseum.com/audio/550e8400.mp3",
                "duration_seconds": 45,
                "file_size_bytes": 1024000,
                "cached": False,
            }
        }


class StreamEventType(str, Enum):
    """SSE stream event types"""

    CONTENT_START = "content_start"
    CONTENT_CHUNK = "content_chunk"
    CONTENT_COMPLETE = "content_complete"
    AUDIO_START = "audio_start"
    AUDIO_COMPLETE = "audio_complete"
    ERROR = "error"


class StreamEvent(BaseModel):
    """SSE stream event schema"""

    event_type: StreamEventType = Field(..., description="Event type")
    data: Dict[str, Any] = Field(..., description="Event data")
    timestamp: datetime = Field(default_factory=datetime.now, description="Event timestamp")

    class Config:
        json_schema_extra = {
            "example": {
                "event_type": "content_complete",
                "data": {"content": "The Mona Lisa is...", "processing_time_ms": 1200},
                "timestamp": "2024-10-03T12:00:00",
            }
        }


class ExplanationError(BaseModel):
    """Error response schema"""

    error: str = Field(..., description="Error type")
    detail: Optional[str] = Field(None, description="Detailed error message")

    class Config:
        json_schema_extra = {
            "example": {"error": "UnsupportedLanguage", "detail": "Language 'xx' is not supported"}
        }

File: frontend/gomuseum_app/lib/features/explanation/domain/entities/explanation.dart
import 'package:equatable/equatable.dart';

/// 艺术品解释实体（纯业务对象，无依赖）
///
/// 包含AI生成的艺术品解释信息，支持多语言和音频输出。
/// 该实体遵循Clean Architecture原则，不依赖任何框架或外部库（除Equatable）。
class Explanation extends Equatable {
  /// 解释记录的唯一标识符
  final String id;

  /// 艺术品名称
  final String artworkName;

  /// 语言代码（en, fr, de, es, it, zh）
  final String language;

  /// 解释文本内容
  final String content;

  /// 音频文件URL（可选）
  final String? audioUrl;

  /// 音频时长（秒）
  final int? audioDurationSeconds;

  /// 创建时间戳
  final DateTime timestamp;

  /// 是否来自缓存
  final bool cached;

  /// 处理时长（毫秒）
  final int processingTimeMs;

  /// 元数据信息
  final ExplanationMetadata metadata;

  const Explanation({
    required this.id,
    required this.artworkName,
    required this.language,
    required this.content,
    this.audioUrl,
    this.audioDurationSeconds,
    required this.timestamp,
    required this.cached,
    required this.processingTimeMs,
    required this.metadata,
  });

  @override
  List<Object?> get props => [
        id,
        artworkName,
        language,
        content,
        audioUrl,
        audioDurationSeconds,
        timestamp,
        cached,
        processingTimeMs,
        metadata,
      ];

  @override
  String toString() {
    return 'Explanation(id: $id, artworkName: $artworkName, language: $language, '
        'cached: $cached, processingTimeMs: ${processingTimeMs}ms, '
        'audioUrl: ${audioUrl != null ? "available" : "none"})';
  }

  /// 复制实例并修改部分字段
  Explanation copyWith({
    String? id,
    String? artworkName,
    String? language,
    String? content,
    String? audioUrl,
    int? audioDurationSeconds,
    DateTime? timestamp,
    bool? cached,
    int? processingTimeMs,
    ExplanationMetadata? metadata,
  }) {
    return Explanation(
      id: id ?? this.id,
      artworkName: artworkName ?? this.artworkName,
      language: language ?? this.language,
      content: content ?? this.content,
      audioUrl: audioUrl ?? this.audioUrl,
      audioDurationSeconds: audioDurationSeconds ?? this.audioDurationSeconds,
      timestamp: timestamp ?? this.timestamp,
      cached: cached ?? this.cached,
      processingTimeMs: processingTimeMs ?? this.processingTimeMs,
      metadata: metadata ?? this.metadata,
    );
  }
}

/// 解释元数据
///
/// 包含解释的详细统计信息和配置参数。
class ExplanationMetadata extends Equatable {
  /// 字数统计
  final int wordCount;

  /// 详细程度（brief, standard, detailed）
  final String detailLevel;

  const ExplanationMetadata({
    required this.wordCount,
    required this.detailLevel,
  });

  @override
  List<Object?> get props => [wordCount, detailLevel];

  @override
  String toString() {
    return 'ExplanationMetadata(wordCount: $wordCount, detailLevel: $detailLevel)';
  }

  /// 复制实例并修改部分字段
  ExplanationMetadata copyWith({
    int? wordCount,
    String? detailLevel,
  }) {
    return ExplanationMetadata(
      wordCount: wordCount ?? this.wordCount,
      detailLevel: detailLevel ?? this.detailLevel,
    );
  }
}
